### PR TITLE
Query string

### LINKED ISSUE #24: [Req]: Query String Tool
### Contact Details

_No response_

### Explain the proposed feature

A tool for working with URL query strings (e.g. `https://somesite.com/?querystring=somevalue&anotherquerystring=somethingelse`).

### Explain the possible use cases for the feature

Query strings can be very long and difficult to parse when debugging complex API calls, having a more human readable format to parse and modify for testing can save a lot of time.

### Any other details you would like to provide?

The tool should at the very least convert a query string to a list of key/values and build a query string from a list of key/values.

For example, the URL above would output the following

```txt
querystring=somevalue
anotherquerystring=somethingelse
```

Additional features that are not required might include:
- Converting to alternate formats such as JSON
- Supporting edge cases like `somesite.com/?json={"this": "is JSON"}`
- Anything else you feel might be useful

### PR DESCRIPTION
**Pull request by nosecreek**
## Type of Pull Request

- [ ] Bug fix
- [x] Enhancement

Related Issue #s or links (if any): 
resolves #24 

## Description of Changes
A tool for working with URL query strings. Converts a query string to a list of key/values and builds a query string from a list of key/values.

### CODE REVIEW

File: components/queryString/QueryParamsForm.tsx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                style={{
+                  width: '100%',
+                  textAlign: 'center',
+                  display: 'flex',
+                  height: '100%',
+                  alignItems: 'center',
+                  justifyContent: 'space-around',
+                }}
+              >
+                <button
```
Comment (dlford) [inline components/queryString/QueryParamsForm.tsx:118]: Lets use the MUI `IconButton` component here to keep the styles consistent

File: components/queryString/QueryParamsForm.tsx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                  <Delete />
+                </button>
+              </div>
+            </Grid>
+          ) : (
+            ''
+          )}
+        </React.Fragment>
+      ))}
+      {/* <Button
```
Comment (dlford) [inline components/queryString/QueryParamsForm.tsx:137]: Lets remove the commented out code here

File: components/queryString/QueryParamsForm.tsx [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+}[];
+export const defaultParams = [{ param: '', value: '', key: 'initial' }];
+
+export interface QueryParamsFormProps {
+  queryParams: QueryParams;
+  setQueryParams: (_queryParams: QueryParams) => void;
+  setQueryString: (_queryString: string) => void;
+}
+
+export function getQueryString(
```
Comment (dlford) [inline components/queryString/QueryParamsForm.tsx:19]: Can we protect the URL here if there is one?

e.g. if I type `https://google.com/?q=test`, then add a new parameter from the inputs it wipes out the URL and I get `q=test&param=value`.

File: components/queryString/QueryParamsForm.tsx [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      spacing={1}
+    >
+      {queryParams.map((param, i) => (
+        <React.Fragment key={param.key}>
+          <Grid
+            item
+            xs={5}
+          >
+            <TextField
+              label={`Parameter ${i + 1}`}
```
Comment (dlford) [inline components/queryString/QueryParamsForm.tsx:84]: Can we add a translation field for this?

File: components/queryString/QueryParamsForm.tsx [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+              onChange={(event) => handleChange(event, i, 'param')}
+              style={{ width: '100%' }}
+            />
+          </Grid>
+          <Grid
+            item
+            xs={6}
+          >
+            <TextField
+              label={`Value ${i + 1}`}
```
Comment (dlford) [inline components/queryString/QueryParamsForm.tsx:96]: Translation field here too if possible

File: components/queryString/QueryParamsForm.tsx [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+              >
+                <button
+                  type='button'
+                  onClick={(event) => deleteRow(event, i)}
+                  style={{
+                    cursor: 'pointer',
+                    background: 'none',
+                    color: 'inherit',
+                  }}
+                  aria-label={`Delete Row ${i + 1}`}
```
Comment (dlford) [inline components/queryString/QueryParamsForm.tsx:126]: Translation field here too if possible

File: components/queryString/QueryStringForm.tsx [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            {t('common:clear')}
+          </Button>
+          <Button
+            startIcon={<ContentCopyIcon />}
+            disabled={!queryString}
+            name='Copy'
+            onClick={() => {
+              navigator.clipboard.writeText(queryString || '').then(
+                () => {
+                  setToastMessage('Copied to clipboard');
```
Comment (dlford) [inline components/queryString/QueryStringForm.tsx:105]: Please the common translation fields for the toast messages, there is one for "copied to clipboard" and "failed to copy"

File: pages/query-string.tsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  });
+
+  const [error, setError] = React.useState('');
+  const [toastOpen, setToastOpen] = useState<boolean>(false);
+  const [toastMessage, setToastMessage] = useState<string>('');
+  const [toastSeverity, setToastSeverity] =
+    useState<ToastProps['severity']>('success');
+  const { t } = useTranslation('queryString');
+  return (
+    <Layout title='Query String'>
```
Comment (dlford) [inline pages/query-string.tsx:34]: Please use the translation field from common here too, same one as the nav item.

File: __TESTS__/query-string.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,164 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+/* eslint-enable import/no-extraneous-dependencies */
+import React from 'react';
+
+import QueryString from '../pages/query-string';
+
+jest.mock('react-i18next', () => ({
```
Comment (dlford) [inline __TESTS__/query-string.spec.tsx:9]: If we add translations for the input field, you will probably need to remove this mock and use `renderWithI18n` from `__TESTS__/helper` instead of `render` from `@testing-library/react`.

Working on a better way to handle this if you have any ideas (that would be a separate PR if you wanted to take it on)

**Review by dlford — CHANGES_REQUESTED** [review]
@nosecreek great job on this! Just a few tweaks needed (it looks like a lot but it's mostly translations).

File: components/queryString/QueryParamsForm.tsx (hunk 3 above)
Comment (nosecreek) [reply components/queryString/QueryParamsForm.tsx:21]: Good idea. I've pushed a commit that solves this by checking if the existing string is a valid URL and then using the JS URL object to update the params. Downside is that it still wipes the URL if it isn't technically valid (i.e. if something like http:// is missing).
Another option would be to split the string at '?' and just keep the first part no matter what it is. I think that would probably work fine, but I'm not sure if there are any edge cases where that might cause problems? Let me know if you have any thoughts.

I'll do the translation stuff later today.

Comment (dlford) [reply components/queryString/QueryParamsForm.tsx:21]: @nosecreek I think splitting at '?' is a good route here, I did a little exploration and at looks like the first '?' is treated as the start of a query and any subsequent '?' is treated as part of the parameter or value depending on which side of the '=' it is on.

Based on that, I think something like this should cover any edge cases I can think of:

```js
const input = 'https://google.com/?q=test?&thing?=value'

const parts = input.split('?');
const url = parts[0];
const query = parts.slice(1).join('?');

console.log(url, query); // "https://google.com/" "?q=test?&thing?=value"
```

Also, thanks for adding a test case for the URL!

Comment (nosecreek) [reply components/queryString/QueryParamsForm.tsx:21]: Makes sense, I changed it to use split instead.

File: __TESTS__/query-string.spec.tsx (hunk 9 above)
Comment (nosecreek) [reply __TESTS__/query-string.spec.tsx:9]: My solution was to use the translation field instead of the English text (i.e. `screen.getByRole('textbox', { name: 'queryString:parameter 1' })` instead of `screen.getByRole('textbox', { name: 'Parameter 1' })`)
Not sure if that's a better solution or not, but I think one advantage is that it works even if the translated text gets changed later? Let me know what you think, I can switch to using renderWithI18n instead if you'd like so that there's better consistency.

Comment (dlford) [reply __TESTS__/query-string.spec.tsx:9]: Yeah, I thought about that, my only concern in doing it that way is if the translation slug is wrong the test will still pass while it is technically not working as expected. But, now that you mention it, TypeScript will catch the missing translation slug, so let's leave it as is, I'll update the other tests and move the mock to `jest.config.js`. Thanks!

File: components/queryString/QueryParamsForm.tsx (hunk 3 above)
Comment (dlford) [reply components/queryString/QueryParamsForm.tsx:22]: We're close on this, but it doesn't quite handle having '?' in a parameter or value, e.g. `https://google.com/?q=test?&thing?=four`

**Review by dlford — APPROVED** [review]
Great job, thank you!
